Can't build with tree-sitter-cli

Hey :wave: 

luarocks-build-treesitter-parser version 5 uses `tree-sitter-cli` to build parsers (`tree-sitter build`)
Using it to build tree-sitter-norg now fails with lots of:

```
/build/source/src/parser.c:634931:1: error: designator order for field 'TSParseAction::<unnamed struct>::child_count
' does not match declaration order in 'TSParseAction::<unnamed struct>'
```

and

```
src/parser.c:469936:1: sorry, unimplemented: non-trivial designated initializers not supported
```

Note: `tree-sitter test` fails with the same errors.

--- a/.envrc
+++ b/.envrc
@@ -1,0 +1,1 @@
+use flake . -Lv

--- a/.github/rockspec.template
+++ b/.github/rockspec.template
@@ -20,6 +20,10 @@ dependencies = $dependencies
 
 test_dependencies = $test_dependencies
 
+build_dependencies = {
+  'luarocks-build-treesitter-parser-cpp',
+}
+
 source = {
   url = repo_url .. '/archive/' .. git_ref .. '.zip',
   dir = '$repo_name-' .. '$archive_dir_suffix',
@@ -32,7 +36,7 @@ if modrev == 'scm' or modrev == 'dev' then
 end
 
 build = {
-  type = "treesitter-parser",
+  type = "treesitter-parser-cpp",
   lang = "norg",
   sources = { "src/parser.c", "src/scanner.cc" },
   copy_directories = $copy_directories,

--- a/.github/workflows/luarocks.yml
+++ b/.github/workflows/luarocks.yml
@@ -1,10 +1,10 @@
 name: Push to Luarocks
 
 on:
+  release:
+    types:
+      - created
   push:
-    release:
-      types:
-        - created
     tags:
       - '*'
   workflow_dispatch:
@@ -28,11 +28,11 @@ jobs:
           luaVersion: "5.1"
       - name: Install Luarocks
         uses: hishamhm/gh-actions-luarocks@master
-      - name: Install `luarocks-build-treesitter-parser` Package
+      - name: Install `luarocks-build-treesitter-parser-cpp` Package
         run: |
-          luarocks --verbose --local --lua-version=5.1 install luarocks-build-treesitter-parser
+          luarocks --verbose --local --lua-version=5.1 install luarocks-build-treesitter-parser-cpp
       - name: LuaRocks Upload
-        uses: nvim-neorocks/luarocks-tag-release@v5
+        uses: nvim-neorocks/luarocks-tag-release@v7
         env:
           LUAROCKS_API_KEY: ${{ secrets.LUAROCKS_API_KEY }}
         with:

--- a/.github/workflows/tree-sitter-norg.rockspec.template
+++ b/.github/workflows/tree-sitter-norg.rockspec.template
@@ -16,7 +16,7 @@ description = {
 }
 
 build_dependencies = {
-  'luarocks-build-treesitter-parser >= 1.5.0',
+  'luarocks-build-treesitter-parser-cpp ~> 2',
 }
 
 source = {
@@ -27,16 +27,5 @@ source = {
 build = {
   type = "treesitter-parser",
   lang = "norg",
-  platforms = {
-    macosx = {
-      libflags = {
-        "-bundle",
-        "-undefined",
-        "dynamic_lookup",
-        "-all_load",
-        "-std=c++11",
-      },
-    },
-  },
   sources = { "src/parser.c", "src/scanner.cc" },
 }

--- a/.gitignore
+++ b/.gitignore
@@ -6,6 +6,7 @@ Cargo.lock
 package-lock.json
 yarn.lock
 *.log
-result/
+result*
 *.wasm
 *.o
+.direnv

--- a/flake.nix
+++ b/flake.nix
@@ -1,31 +1,48 @@
 {
+  description = "tree-sitter parser for Neorg";
+
   inputs = {
-    nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
-    flake-utils.url = "github:numtide/flake-utils";
-    flake-compat = {
-      url = "github:edolstra/flake-compat";
-      flake = false;
-    };
+    nixpkgs.url = "github:NixOS/nixpkgs";
+    flake-parts.url = "github:hercules-ci/flake-parts";
   };
-  outputs = { self, nixpkgs, flake-utils, ... }@inputs:
-    let
-      defaultPackage = pkgs: pkgs.callPackage (nixpkgs + "/pkgs/development/tools/parsing/tree-sitter/grammar.nix") { } {
-        language = "norg";
-        source = ./.;
-        inherit (pkgs.tree-sitter) version;
+
+  outputs = inputs @ {
+    self,
+    nixpkgs,
+    flake-parts,
+    ...
+  }:
+    flake-parts.lib.mkFlake {inherit inputs;} {
+      systems = [
+        "x86_64-linux"
+        "x86_64-darwin"
+        "aarch64-darwin"
+      ];
+      perSystem = {system, ...}: let
+        pkgs = import nixpkgs {
+          inherit system;
+          overlays = [
+            self.overlays.default
+          ];
+        };
+      in {
+        devShells.default = pkgs.mkShell {
+          name = "tree-sitter-norg devShell";
+          buildInputs = with pkgs;
+          with pkgs; [
+            nil
+            alejandra
+            clang
+            tree-sitter
+          ];
+        };
+        packages = rec {
+          default = tree-sitter-norg;
+          inherit (pkgs.lua51Packages) tree-sitter-norg;
+        };
       };
-    in
-    (flake-utils.lib.eachDefaultSystem
-      (system:
-        let pkgs = import nixpkgs { inherit system; }; in
-        {
-          defaultPackage = defaultPackage pkgs;
-          devShell = pkgs.mkShell {
-            nativeBuildInputs = with pkgs; [
-              nodejs
-              nodePackages.node-gyp
-              tree-sitter
-            ];
-          };
-        })) // (let pkgs = import nixpkgs { }; in { defaultPackage = defaultPackage pkgs; });
+      flake = {
+        overlays.default = import ./nix/overlay.nix {inherit self;};
+      };
+    };
 }

--- a/nix/overlay.nix
+++ b/nix/overlay.nix
@@ -1,0 +1,23 @@
+{self}: final: prev: let
+  luaPackage-override = luaself: luaprev: {
+    tree-sitter-norg = luaprev.tree-sitter-norg.overrideAttrs (oa: {
+      version = "scm-1";
+      knownRockspec = "${self}/tree-sitter-norg-scm-1.rockspec";
+      src = self;
+      propagatedBuildInputs = [
+        # nixpkgs still uses luarocks-build-treesitter-parser (not the -cpp one).
+        luaself.luarocks-build-treesitter-parser-cpp
+      ];
+    });
+  };
+in {
+  lua5_1 = prev.lua5_1.override {
+    packageOverrides = luaPackage-override;
+  };
+  lua51Packages = prev.lua51Packages // final.lua5_1.pkgs;
+  vimPlugins =
+    prev.vimPlugins
+    // {
+      tree-sitter-norg = final.neovimUtils.buildNeovimPlugin {luaAttr = "tree-sitter-norg";};
+    };
+}

--- a/tree-sitter-norg-scm-1.rockspec
+++ b/tree-sitter-norg-scm-1.rockspec
@@ -2,6 +2,8 @@ package = "tree-sitter-norg"
 
 version = "scm-1"
 
+rockspec_format = "3.0"
+
 source = {
   url = "git://github.com/nvim-neorg/tree-sitter-norg",
 }
@@ -14,11 +16,14 @@ description = {
 
 dependencies = {
   "lua >= 5.1",
-  "luarocks-build-treesitter-parser >= 1.2.0",
+}
+
+build_dependencies = {
+  "luarocks-build-treesitter-parser-cpp ~> 2",
 }
 
 build = {
-  type = "treesitter-parser",
+  type = "treesitter-parser-cpp",
   lang = "norg",
   sources = { "src/parser.c", "src/scanner.cc" },
 }

